### PR TITLE
[Model Compression] Pruning Wrapper Refactor

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
Refactor the pruner wrapper to preserve masked weight values ​​during pruning.

### Checklist ###
  - [ ] test case
  - [ ] doc

### How to test ###
Run any v2 pruners and check if they can work. Suggest `TaylorFOWeightPruner`.

### CODE REVIEW

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 1]
```diff
@@ -38,56 +38,12 @@ class PrunerScoredModuleWrapper(Module):
...
         self.weight_score = Parameter(torch.empty(self.weight.size()))
         torch.nn.init.constant_(self.weight_score, val=0.0)
 
-        # register buffer for mask
-        self.register_buffer("weight_mask", torch.ones(self.module.weight.shape))
-        if hasattr(self.module, 'bias') and self.module.bias is not None:
-            self.register_buffer("bias_mask", torch.ones(self.module.bias.shape))
-            self.bias = Parameter(torch.empty(self.module.bias.size()))
-        else:
-            self.register_buffer("bias_mask", None)
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:63]: so where are masks are generated then?

Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:63]: all these logic are moved to its parent's class `PrunerModuleWrapper`

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 2]
```diff
@@ -699,7 +699,7 @@ def _calculate_taylor_expansion(self, weight_tensor: Tensor, grad: Tensor) -> Te
         return (weight_tensor.detach() * grad.detach()).data.pow(2)
 
     def reset_tools(self):
-        hook_targets = {layer_info.name: layer_info.module.weight for layer_info, _ in self._detect_modules_to_compress()}
+        hook_targets = {name: wrapper.weight for name, wrapper in self.get_modules_wrapper().items()}
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:702]: so, weight as a member variable in wrapper always exists? no matter wrapped or unwrapped?

**Comment by QuanluZhang** [comment]
better to add ut to verify that masked weights are indeed preserved

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:702]: only wrapped wrapper has `weight`, the `weight` will be del when unwrapped.

Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:702]: Yes, the weight will always exists... Sometimes we need to `_unwrap_model()` to get the model graph, then `_wrap_model()`, so reserving `weight` after unwrapped is necessary.

**Comment by J-shang** [comment]
> @J-shang better to add ut to verify that masked weights are indeed preserved

Good suggestion, I will add it.

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 1 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:63]: got it

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:702]: got it

**Comment by J-shang** [comment]
> @J-shang better to add ut to verify that masked weights are indeed preserved

add it

**Review by QuanluZhang — APPROVED** [review]

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 2 above)
Comment (zheng-ningxin) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:702]: I'm just curious, what's the specific benefit of this(allocating the weight and bias as  the member variable) compared to the previous way? Performance or cost less memory?

**Review by zheng-ningxin — APPROVED** [review]

**Comment by J-shang** [comment]
This pr will merge after doc refactor.
